### PR TITLE
[lld][ELF] Demote symbols in discarded sections to Undefined.

### PR DESCRIPTION
**Pull request by bevin-hansson**
After linker script processing, it may be the case that sections
were discarded via /DISCARD/, but relocations to symbols in those
sections remain. Currently, such symbols are still considered
Defined even though their section is not live and will not be
placed anywhere. This results in a silently broken binary.

This patch goes through the symbols after placement and changes
them from Defined to Undefined if their section is no longer
live at that point. During relocation processing, we will catch
such undefined symbols and report an error.

See #58891.

### CODE REVIEW

File: lld/ELF/Writer.cpp
Comment (bevin-hansson) [inline lld/ELF/Writer.cpp:720]: I'm a bit unsure about this. I had to prevent this for section symbols, because a few tests fail otherwise, mostly ones related to arm exidx sections. I don't understand the mechanisms at play enough to know exactly what is supposed to happen.

File: lld/test/ELF/linkerscript/discard-symbol.s
Comment (smithp35) [inline lld/test/ELF/linkerscript/discard-symbol.s:1]: The test looks only to check global symbols. The patch makes modifications for local symbols and symbol ordering files, please can you add test cases for these as well?

Will be worth adding a test case for relocatable links too.

File: lld/ELF/Driver.cpp
Comment (smithp35) [inline lld/ELF/Driver.cpp:3069]: Could it be worth adding a script->seenDiscard variable so that this potentially expensive operation is only done when someone has used /DISCARD/ ?

File: lld/ELF/Writer.cpp
Comment (smithp35) [reply lld/ELF/Writer.cpp:720]: I'm surprised about test failures in .ARM.exidx. As there shouldn't be any relocations to a .ARM.exidx input section. There will be relocations from a .ARM.exidx section, but given that when a section is discarded, any section with a SHF_LINK_ORDER dependency on it is discarded.

**Review by smithp35 — COMMENTED** [review]
Does GNU ld permit this, or does it result in dangling relocations? If GNU ld does permit this then there's a chance that this might break some programs where the dangling relocations are benign. I guess that we may need to add a command line option to suppress the errors, although maybe not initially.

How does this work with uses of /DISCARD/ in a ld -r relocatable link? I guess a follow up question is what is it supposed to do for ld -r relocatable links. It would still be user error to /DISCARD/ a section where some other section had relocations to that section, but I don't think undefined symbols are checked for.

A possible alternative implementation could be to test each relocation in scanOne to see if the symbol is defined in a discarded section. You could then make a specific error message, or do the work to make the symbol undefined at that point. Not sure how well that would work out, but could be worth a prototype.

File: lld/test/ELF/linkerscript/discard-symbol.s
Comment (bevin-hansson) [reply lld/test/ELF/linkerscript/discard-symbol.s:1]: This is true. The reason I couldn't add a test for the local symbol case is because when you have local symbols, llvm-mc seems to emit them as section symbols with an offset addend rather than an actual symbol relocation. So it runs afoul of the exception for section symbols, and doesn't trigger.

File: lld/ELF/Writer.cpp
Comment (bevin-hansson) [reply lld/ELF/Writer.cpp:720]: I can disable the check and push so you can see what tests are failing on it.

**Comment by bevin-hansson** [comment]
> Does GNU ld permit this, or does it result in dangling relocations? If GNU ld does permit this then there's a chance that this might break some programs where the dangling relocations are benign. I guess that we may need to add a command line option to suppress the errors, although maybe not initially.

When I tested a while back, GNU ld seems to keep the section if there is still a reference to it, even though it was marked /DISCARD/. I don't think this is possible in lld, since the relocation processing is done after the linker script handling, and we've already made the call to keep or not keep the section at that point.

> How does this work with uses of /DISCARD/ in a ld -r relocatable link? I guess a follow up question is what is it supposed to do for ld -r relocatable links.

I haven't tested -r. I would assume that discarding something in an -r invocation today does not trigger these types of errors, so they wouldn't with this patch either.

> It would still be user error to /DISCARD/ a section where some other section had relocations to that section, but I don't think undefined symbols are checked for.

Hm, would it? Could you not discard a section that contains something which is referenced from the current file, and have that reference be resolved from some other object at the final link?

> A possible alternative implementation could be to test each relocation in scanOne to see if the symbol is defined in a discarded section. You could then make a specific error message, or do the work to make the symbol undefined at that point. Not sure how well that would work out, but could be worth a prototype.

Yeah, it doesn't have to go through this logic. It's just that the logic for emitting these errors already existed (and only worked for Undefined), so I went that path.

**Comment by smithp35** [comment]
> > Does GNU ld permit this, or does it result in dangling relocations? If GNU ld does permit this then there's a chance that this might break some programs where the dangling relocations are benign. I guess that we may need to add a command line option to suppress the errors, although maybe not initially.
> 
> When I tested a while back, GNU ld seems to keep the section if there is still a reference to it, even though it was marked /DISCARD/. I don't think this is possible in lld, since the relocation processing is done after the linker script handling, and we've already made the call to keep or not keep the section at that point.
>

I had a quick check and GNU ld on my machine at least gives a similar error message to what you are proposing.
 
> > How does this work with uses of /DISCARD/ in a ld -r relocatable link? I guess a follow up question is what is it supposed to do for ld -r relocatable links.
> 
> I haven't tested -r. I would assume that discarding something in an -r invocation today does not trigger these types of errors, so they wouldn't with this patch either.
> 

It looks like GNU ld does trigger these errors with ld -r at least with a trivial test case. LLD seems to give a warning in this case and outputs a R_*_NONE relocation in its place.

> > It would still be user error to /DISCARD/ a section where some other section had relocations to that section, but I don't think undefined symbols are checked for.
> 
> Hm, would it? Could you not discard a section that contains something which is referenced from the current file, and have that reference be resolved from some other object at the final link?
> 
I think it depends on whether the linker resolves the relocation at relocatable link time, and whether it is a global definition or not. For example if we make a local symbol undefined that is likely to be unrecoverable.

> > A possible alternative implementation could be to test each relocation in scanOne to see if the symbol is defined in a discarded section. You could then make a specific error message, or do the work to make the symbol undefined at that point. Not sure how well that would work out, but could be worth a prototype.
> 
> Yeah, it doesn't have to go through this logic. It's just that the logic for emitting these errors already existed (and only worked for Undefined), so I went that path.

I've found out why the tests are failing anyway. The ELF/linkerscript/discard-section.s contains a local reference that we would explicitly want to catch, so I think this test would need updating. The ARM.exidx tests are failing due a missing test for liveness in scanRelocations. The .ARM.exidx SyntheticSection has a list of input .ARM.exidx sections. When some, but not all of these input sections are discarded we need the liveness check.

This should fix that:
```
diff --git a/lld/ELF/Relocations.cpp b/lld/ELF/Relocations.cpp
index 441111616196..f3fb0c71a8b3 100644
--- a/lld/ELF/Relocations.cpp
+++ b/lld/ELF/Relocations.cpp
@@ -1574,7 +1574,8 @@ template <class ELFT> void elf::scanRelocations() {
         scanner.template scanSection<ELFT>(*sec);
       if (part.armExidx && part.armExidx->isLive())
         for (InputSection *sec : part.armExidx->exidxSections)
-          scanner.template scanSection<ELFT>(*sec);
+          if (sec->isLive())
+            scanner.template scanSection<ELFT>(*sec);
     }
   });
 }
```

With this you should be able remove the special case for section symbols. It should be possible to use yaml2obj to make a test case with a relocation to a local symbol, but with the special case removed I don't think you'll need them.

**Comment by bevin-hansson** [comment]
> > > Does GNU ld permit this, or does it result in dangling relocations? If GNU ld does permit this then there's a chance that this might break some programs where the dangling relocations are benign. I guess that we may need to add a command line option to suppress the errors, although maybe not initially.
> > 
> > 
> > When I tested a while back, GNU ld seems to keep the section if there is still a reference to it, even though it was marked /DISCARD/. I don't think this is possible in lld, since the relocation processing is done after the linker script handling, and we've already made the call to keep or not keep the section at that point.
> 
> I had a quick check and GNU ld on my machine at least gives a similar error message to what you are proposing.

Interesting. I'll have to try it again myself.

> > > How does this work with uses of /DISCARD/ in a ld -r relocatable link? I guess a follow up question is what is it supposed to do for ld -r relocatable links.
> > 
> > 
> > I haven't tested -r. I would assume that discarding something in an -r invocation today does not trigger these types of errors, so they wouldn't with this patch either.
> 
> It looks like GNU ld does trigger these errors with ld -r at least with a trivial test case. LLD seems to give a warning in this case and outputs a R_*_NONE relocation in its place.
> 
> > > It would still be user error to /DISCARD/ a section where some other section had relocations to that section, but I don't think undefined symbols are checked for.
> > 
> > 
> > Hm, would it? Could you not discard a section that contains something which is referenced from the current file, and have that reference be resolved from some other object at the final link?
> 
> I think it depends on whether the linker resolves the relocation at relocatable link time, and whether it is a global definition or not. For example if we make a local symbol undefined that is likely to be unrecoverable.

Ah, that's true. Unresolved locals would be a problem.

> > > A possible alternative implementation could be to test each relocation in scanOne to see if the symbol is defined in a discarded section. You could then make a specific error message, or do the work to make the symbol undefined at that point. Not sure how well that would work out, but could be worth a prototype.
> > 
> > 
> > Yeah, it doesn't have to go through this logic. It's just that the logic for emitting these errors already existed (and only worked for Undefined), so I went that path.
> 
> I've found out why the tests are failing anyway. The ELF/linkerscript/discard-section.s contains a local reference that we would explicitly want to catch, so I think this test would need updating. The ARM.exidx tests are failing due a missing test for liveness in scanRelocations. The .ARM.exidx SyntheticSection has a list of input .ARM.exidx sections. When some, but not all of these input sections are discarded we need the liveness check.
> 
> This should fix that:
> 
> ```
> diff --git a/lld/ELF/Relocations.cpp b/lld/ELF/Relocations.cpp
> index 441111616196..f3fb0c71a8b3 100644
> --- a/lld/ELF/Relocations.cpp
> +++ b/lld/ELF/Relocations.cpp
> @@ -1574,7 +1574,8 @@ template <class ELFT> void elf::scanRelocations() {
>          scanner.template scanSection<ELFT>(*sec);
>        if (part.armExidx && part.armExidx->isLive())
>          for (InputSection *sec : part.armExidx->exidxSections)
> -          scanner.template scanSection<ELFT>(*sec);
> +          if (sec->isLive())
> +            scanner.template scanSection<ELFT>(*sec);
>      }
>    });
>  }
> ```
> 
> With this you should be able remove the special case for section symbols. It should be possible to use yaml2obj to make a test case with a relocation to a local symbol, but with the special case removed I don't think you'll need them.

Great, big thanks for finding this! I pushed the fix and test additions, along with some other suggested changes.

**Review by smithp35 — APPROVED** [review]
Thanks very much for the updates. This looks good to me know. I think it is worth the maintainer @MaskRay taking a look first before merging.

File: lld/test/ELF/linkerscript/discard-section-reloc.s
Comment (MaskRay) [inline lld/test/ELF/linkerscript/discard-section-reloc.s:1]: This is copied from discard-section.s but adds `-r`. We should use the same file: https://maskray.me/blog/2021-08-08-toolchain-testing#i-dont-know-an-existing-test-can-be-enhanced

**Review by MaskRay — CHANGES_REQUESTED** [review]

File: lld/test/ELF/linkerscript/discard-section-reloc.s
Comment (MaskRay) [reply lld/test/ELF/linkerscript/discard-section-reloc.s:1]: Please remove the new file discard-symbol.s as well.

You can squash the commits and rebase on top of 4fb49f44fdf558a942de2b0fc81e7f1fdf1c798c

File: lld/ELF/Driver.cpp
Comment (MaskRay) [inline lld/ELF/Driver.cpp:3065]: "Explicitly" is redundant.

Suggest: Demote symbols defined relative to input sections that are discarded by /DISCARD/ so that relocations referencing them will get reported.

File: lld/ELF/Driver.cpp
Comment (MaskRay) [inline lld/ELF/Driver.cpp:3072]: This `if` use case requires `{}`. However, we can use an early return:
```
Defined *d = ...
if (!d)
  continue;
...
```

**Review by MaskRay — CHANGES_REQUESTED** [review]

File: lld/test/ELF/linkerscript/discard-section-reloc.s
Comment (MaskRay) [reply lld/test/ELF/linkerscript/discard-section-reloc.s:1]: Rebase on top of 557299c9b6464f27968904aad5429cfb1512434e

`weakref1` is test case about a symbol defined in a discarded section (`/DISCARD/`) that is weak referenced. It's controversial whether we should report an error. Since our implementation reports an error and GNU ld reports an error as well, reporting an error seems fine.

File: lld/test/ELF/linkerscript/discard-symbol.s
Comment (MaskRay) [reply lld/test/ELF/linkerscript/discard-symbol.s:1]: Remove the file after Rebase on top of https://github.com/llvm/llvm-project/commit/557299c9b6464f27968904aad5429cfb1512434e

**Comment by bevin-hansson** [comment]
I rebased, adapted the test and addressed the comments.

File: lld/ELF/Driver.cpp
Comment (bevin-hansson) [inline lld/ELF/Driver.cpp:3081]: I changed the section index code in both places. It was not giving the right index. I'm not sure why I didn't notice it earlier. I'm pretty sure it worked before.

File: lld/ELF/Driver.cpp
Comment (MaskRay) [inline lld/ELF/Driver.cpp:3066]: It's a corner case but I think is worth testing. The demotion logic applies `--gc-sections` as well but we only run the code when `/DISCARD/` is seen, so there will be some behavior differences.

I have updated `lld/test/ELF/gc-sections-tls.s` in https://github.com/MaskRay/llvm-project/tree/lld-symbols-in-discard  to test this.

Actually, we probably should do #69295 so that all non-local symbols in discarded sections are demoted, regardless of whether `/DISCARD/` is seen? @smithp35

**Comment by bevin-hansson** [comment]
Your alternate approach seems good, I was also worried about the performance impact of looking through all the symbols like that.

The TLS case was a bit unexpected to me. I wouldn't have thought that the demotion could cause any issues related to --gc-sections, since sections with references to them cannot be discarded by it.

**Comment by MaskRay** [comment]
Superseded by #69295.

**Pull request closed without merging**
